tensorflow_addons/optimizers/discriminative_layer_training_test.py

```diff
@@ -0,0 +1,467 @@
+
+    def _test_config_tofrom(self, model_fn, loss, opt):
+        """Confirm that optimizer saves config and loads config."""
+
+        # build model and save the opt to a config as c.
+        learning_rate = 0.01
+        model_lr = model_fn()
+        model_lr.layers[0].lr_mult = 0.3
+        model_lr.layers[0].layers[-1].lr_mult = 0.1
+        model_lr.layers[-1].lr_mult = 0.5
+
+        d_opt = DiscriminativeLayerOptimizer(
+            opt, model_lr, verbose=False, learning_rate=learning_rate
+        )
+        model_lr.compile(loss=loss, optimizer=d_opt)
+
+        c = d_opt.get_config()
+
+        # reconstruct the model and then build the opt from config.
+
+        model_lr = model_fn()
+        model_lr.layers[0].lr_mult = 0.3
+        model_lr.layers[0].layers[-1].lr_mult = 0.1
+        model_lr.layers[-1].lr_mult = 0.5
+
+        d_opt_from_config = DiscriminativeLayerOptimizer.from_config(c, model_lr)
+        model_lr.compile(loss=loss, optimizer=d_opt_from_config)
+
+        # we expect both optimizers to have the same optimizer group and base optimizer.
+        self.assertAllEqual(
+            len(d_opt.optimizer_group), len(d_opt_from_config.optimizer_group)
+        )
+        self.assertAllEqual(d_opt.opt_class, d_opt_from_config.opt_class)
+
+        # we also expect the lr for each opt in the opt groups to be the same. Also confirms same lr mult.
+        self.assertAllEqual(
+            [opt.learning_rate for opt in d_opt.optimizer_group],
+            [opt.learning_rate for opt in d_opt_from_config.optimizer_group],
+        )
+
+
+def wrap_test(func, **kwargs):
+    """Wrap the test method so that it has pre assigned kwargs."""
+
+    def test(self):
+        return func(self, **kwargs)
+
+    return test
+
+
+def generate_tests():
+    # Generate tests for each permutation in the zipped permutes.
+    # This separates tests for each permuatation of model, optimizer, and loss.
+    for name, func in DiscriminativeLearningTest.__dict__.copy().items():
+        if callable(func) and name[:5] == "_test":
+            for model_fn, loss, opt in _zipped_permutes():
+
+                # Name the test as test_testname_model_loss_optimizer.
+                testmethodname = name[1:] + "_%s_%s_%s" % (
+                    model_fn.__name__,
+                    loss.name,
+                    opt.__name__,
+                )
+
+                # Create test functions that use kwargs mentioned above.
+                testmethod_dist = wrap_test(
+                    func=func, model_fn=model_fn, loss=loss, opt=opt,
+                )
+
+                # Set class attributes so we get multiple nicely named tests.
+                # Also all tests are set to run distributed, so append distributed to the end.
+                setattr(
+                    DiscriminativeLearningTest, testmethodname, testmethod_dist,
+                )
+
+
+if __name__ == "__main__":
+    generate_tests()
+    tf.test.main()
```

Comment:
```suggestion
if __name__ == "__main__":
    sys.exit(pytest.main([__file__]))
```

We're moving to pytest to make the testing easier.